### ISSUE TITLE
Official headless ARM support

### ISSUE THREAD

**Issue by jae1911** [Yellow-Dog-Man/Resonite-Issues#2555]
### Is your feature request related to a problem? Please describe.

Currently, it is not possible to run the headless on ARM CPUs without a bit of tinkering.

The architecture is also officially unsupported.

### Describe the solution you'd like

As ARM-based servers are becoming increasingly popular with cloud providers (AWS, Oracle, GCP, Azure, Hetzner), it would be nice to have official support for the ARM architecture to host headlesses.

The Oracle Cloud Free Tier is also a popular way for people to get a server to host anything they want, including headlesses.
Adding an official ARM build would make it more accessible and easier to host headlesses in general.

### Describe alternatives you've considered

Continue with the tinkering.

### Additional Context

See Discord thread at: https://discord.com/channels/1040316820650991766/1158936045186973726

### Requesters

U-j4 | j4.lc (Discord)
@gameboycjp is also very active on this subject on the Discord

**Comment by jae1911** [comment]
> We can put a list together for which ones would be needed.

That would be helpful!

**Comment by gameboycjp** [comment]
FreeImage, FreeType, and Opus are the minimum.

**Comment by gameboycjp** [comment]
There's also Steamworks.NET targeting x86/64 exclusively and causing an error when reflection is used. Setting it to build for AnyCPU fixes the problem.

**Comment by ticpu** [comment]
What about making Steamworks optional for the headless?  It isn't required for anything useful that I know of when you setup your port forwards and all.

**Comment by gameboycjp** [comment]
I'm not sure that's necessary. Once Steamworks.NET is built for AnyCPU, the headless works as it does on mono. I think I'll put together a simple PR, I just need to figure out the best approach for the change.

**Comment by Frooxius** [comment]
I'm a bit puzzled on how would Steamworks work on ARM? It's a wrapper for the native Steam library and that one is for x64 as far as I'm aware.

Does Steam offer ARM version of it?

**Comment by Raidriar796** [comment]
Steam does not officially provide support for ARM, but there are many projects and some package managers that provide ARM versions of SteamCMD. DepotDownloader also works on ARM but it isn't as good for how we use headless software as it's more built for just downloading the latest or a specific version of a software from Steam instead of updating an existing install.

**Comment by gameboycjp** [comment]
> Does Steam offer ARM version of it?

It doesn't, but the wrapper needs to be able to be reflected against during engine initialization. I can provide more info here or in #2545, but fixing that is a prerequisite for getting the .NET 8 headless to load on ARM.

**Comment by ticpu** [comment]
I'm not sure why my comment hasn't been considered, on the mono version, it currently fails to load and continues loading the worldson my ARM instance. What is the use of Steamworks on the headless in the first place?  Can it just be made as an optional dependency?

Logs for my current ARM instance:
```
17:06:38.039            Steam
17:06:38.075    Exception initializing Steamworks.NET!
System.DllNotFoundException: steam_api assembly:<unknown assembly> type:<unknown type> member:(null)
<stack trace continues...>
at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00000] in <12b418a7818c4ca0893feeaaf67f1e7f>:0
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <12b418a7818c4ca0893feeaaf67f1e7f>:0
17:06:38.090    Supported network protocols: lnl-nat, lnl
17:06:38.092    Initializing SkyFrost Interface...
17:06:38.135    Proxy configuration for skyfrost not found or failed to initialize.
```

**Comment by gameboycjp** [comment]
Yeah, rebuilding Steamworks.NET for AnyCPU brings .NET8 in line with Mono on ARM. I think it's type scanning that fails, judging from the logs. I also noticed the cache never gets the metadata for frooxengine stored.

**Comment by Frooxius** [comment]
Steamworks is used for Steam Networking Sockets protocol. Without it, this protocol won't be available.

It's not going to make the rest not work, so it should work fine regardless. Ideally it should just fail to load and continue without it, so fixing up that would be the first step for that.

**Comment by charlie-sans** [comment]
hey! how's progress on this issue going. just wanting to see if there is anything that i can look into for assistance for yall

**Comment by ProbablePrime** [comment]
> hey! how's progress on this issue going. just wanting to see if there is anything that i can look into for assistance for yall

No one is actively working on this right now.

You can see what we need here: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2555#issuecomment-2225930749

**Comment by gameboycjp** [comment]
> The main thing we need for this are ARM versions of the core libraries we use. We do have repositories for these open, so it's possible for community to contribute CI/CD scripts that would get those versions built which would help this.

I've been working on these, and I finally got around to asking where some of the ones that weren't forked by y'all were in Primes office hour today. Where would I contribute those CI scripts to, for the libraries that aren't forked?

**Comment by Frooxius** [comment]
> > The main thing we need for this are ARM versions of the core libraries we use. We do have repositories for these open, so it's possible for community to contribute CI/CD scripts that would get those versions built which would help this.
> 
> I've been working on these, and I finally got around to asking where some of the ones that weren't forked by y'all were in Primes office hour today. Where would I contribute those CI scripts to, for the libraries that aren't forked?

Make a request for us to put up forks for any that are missing and we'll sort them out!

**Comment by gameboycjp** [comment]
Opus, Brotli, and FreeType, currently.

**Comment by ticpu** [comment]
Please also consider a single repo than uses submodules where a fork isn't necessary as to keep things in line with upstream as much as possible. I don't see the day where libopus will have to be customized for Resonite.

I've been using my system's own library for most of the dependencies for the headless, even the forked one for some, and it just seemed to work. And it is even better as the distro keep them updated for security versus a fork.

**Comment by Frooxius** [comment]
Even if we might not require modifications, we usually fork so we can verify the changes coming into Resonite and make sure they don't break things (which has happened a few times by just updating a dependency).

**Comment by gameboycjp** [comment]
Two I've missed are Compressonator and PDFium.

**Comment by Frooxius** [comment]
Both of these we are just using the prebuilt ones.

Compressonator is not supported on Linux right now, we'd need to setup pipeline to build shared libraries for those.

**Comment by jae1911** [comment]
Just as an index for current PRs for ARM support:

- Opus: https://github.com/Yellow-Dog-Man/opus/pull/2
- FreeImage: https://github.com/Yellow-Dog-Man/FreeImage/pull/18

I'll update this comment as more pile up.

**Comment by jae1911** [comment]
Alright, I updated the workflows PRs for those two repos to use the new ARM runners.

FreeImage: https://github.com/Yellow-Dog-Man/FreeImage/pull/18#issuecomment-2597803552
Opus: https://github.com/Yellow-Dog-Man/opus/pull/3

Anybody could make a list of the remaining libraries to build that way? I think now it will be really fast to get those up and working.

**Comment by jae1911** [comment]
Another PR (this time just updated) that will come in handy later when #880 lands.

LibHeif: https://github.com/Yellow-Dog-Man/libheif/pull/1

**Comment by jae1911** [comment]
Given I'm hammering down the CI/CD for missing libs, any way to get this assigned?

**Comment by ProbablePrime** [comment]
Done :)

**Comment by jae1911** [comment]
It's me again, coming with some gifts for this first day of MMC.

New PRs opened today:
- Crunch: https://github.com/Yellow-Dog-Man/crunch/pull/3
- Assimp: https://github.com/Yellow-Dog-Man/assimp/pull/4
- MSDFGen: https://github.com/Yellow-Dog-Man/msdfgen/pull/2 (couldn't get the Windows build working on this one, it's explained why in the PR)
- RNNoise: https://github.com/Yellow-Dog-Man/rnnoise/pull/3

Currently missing are:
- Freetype which has no fork in the YDMS org (at least not publicly viewable)
- Brotli (just to confirm, it is https://github.com/Yellow-Dog-Man/brotli.net this one right?)

**Comment by Frooxius** [comment]
I switched it to progress. I haven't had change to look at the PR's and publishing other repos yet, sorry!

**Comment by jae1911** [comment]
New PR on the new brotli repo: https://github.com/Yellow-Dog-Man/brotli/pull/1

This one needs some testing as the version is quite newer than what was previously built and used. But I bet it would still work.

**Comment by jae1911** [comment]
Doing a quick ping here about a possible FreeType fork which is now the only one missing to reach full ARM support. 

CC: @ProbablePrime (from office hours)

**Comment by jae1911** [comment]
New PR: compressonator. Not sure if this is needed for the headless, but the patches contained in the PR might help later for #1412 which I plan on tackling after this issue.

https://github.com/Yellow-Dog-Man/compressonator/pull/6

**Comment by jae1911** [comment]
New PR: SoundPipe, https://github.com/Yellow-Dog-Man/soundpipe/pull/1

To have full compatibility, we'd still need two forks:
- FreeType: https://gitlab.freedesktop.org/freetype/freetype
- SteamAudio: https://github.com/ValveSoftware/steam-audio
